File: Problems/5_Longest_Palindromic_Substring.py
class Solution:
	def longestPalindrome(self, s):
		maxLen=1
		start=0
		# ans = ''
		for i in range(len(s)):
			if i-maxLen >=1 and s[i-maxLen-1:i+1]==s[i-maxLen-1:i+1][::-1]:
				start=i-maxLen-1
				maxLen+=2
				# ans = s[start:start+maxLen]
				continue

			if i-maxLen >=0 and s[i-maxLen:i+1]==s[i-maxLen:i+1][::-1]:
				start=i-maxLen
				maxLen+=1
				# ans = s[start:start+maxLen]
		return s[start:start+maxLen]

# class Solution:
#     def longestPalindrome(self, s: str) -> str:
#         self.s = s
#         max_len_left = 0
#         max_len_right = 0
#         max_len = 1
#         i = 0
#         for i in range(len(s)):
#             middle = i
#             left = middle
#             right = middle+1
#             while left>=0 and right<len(s) and s[left]==s[right]:
#                 if right-left+1>max_len:
#                     max_len = right-left+1
#                     max_len_left  = left
#                     max_len_right = right
#                 left  -= 1
#                 right += 1

#             left  = middle-1
#             right = middle+1
#             while left>=0 and right<len(s) and s[left]==s[right]:
#                 if right-left+1>max_len:
#                     max_len = right-left+1
#                     max_len_left  = left
#                     max_len_right = right
#                 left  -= 1
#                 right += 1
#             i += 1        
#         return s[max_len_left:max_len_right+1]

solution = Solution()
print(solution.longestPalindrome('babdasspddggp'))

File: Problems/5_Longest_Palindromic_Substring.cpp
#include<iostream>
#include<string>
#include<algorithm>

using namespace std;

class Solution {
public:
    string longestPalindrome(string s) {
        int maxLen = 1;
        int start  = 0;
        for(int i=0; i<s.length(); ++i){
            if(i-maxLen>=1){
                bool flag = false;
                string sub1(s.substr(i-maxLen-1, maxLen+2));
                string sub2(sub1.rbegin(), sub1.rend());
                if(sub1==sub2){
                    start = i-maxLen-1;
                    maxLen += 2;
                    flag = true;
                }
                if (flag){
                    continue;
                }
            }
            if(i-maxLen>=0){
                string sub1(s.substr(i-maxLen, maxLen+1));
                string sub2(sub1.rbegin(), sub1.rend());
                if(sub1==sub2){
                    start = i-maxLen;
                    maxLen += 1;
                }
            }
        }
        return s.substr(start, maxLen);
    }
};

int main(){
    string s = "cbbd";
    Solution sol;
    cout << sol.longestPalindrome(s);
    return 0;
}